File: sdk/java-guest/runtime/dev/latent/guest/Result.java
package dev.latent.guest;

/** A declared WIT error is a value; arbitrary Java exceptions are never coerced. */
public final class Result<T, E> {
    private final boolean error;
    private final T value;
    private final E failure;
    private Result(boolean error, T value, E failure) {
        this.error = error; this.value = value; this.failure = failure;
    }
    public static <T, E> Result<T, E> ok(T value) {
        return new Result<>(false, java.util.Objects.requireNonNull(value), null);
    }
    public static <T, E> Result<T, E> err(E value) {
        return new Result<>(true, null, java.util.Objects.requireNonNull(value));
    }
    public boolean isError() { return error; }
    public T value() {
        if (error) throw new IllegalStateException("declared error has no success value");
        return value;
    }
    public E error() {
        if (!error) throw new IllegalStateException("success has no declared error");
        return failure;
    }
}

File: sdk/typescript-guest/capabilities/result.ts
/** WIT declared errors are data. An ordinary JavaScript exception still traps. */
export type Result<T, E> = { tag: 'ok'; val: T } | { tag: 'err'; val: E };

export function call<T, E extends { tag: string }>(operation: () => T,
    tags: readonly E['tag'][]): Result<T, E> {
  try { return { tag: 'ok', val: operation() }; }
  catch (error) {
    // The pinned maintained generator puts a WIT result error in Error.payload.
    // Never translate arbitrary exceptions or resource misuse into WIT success.
    // The generated binding and application use distinct SpiderMonkey realms;
    // instanceof uses the local Error prototype and rejects valid WIT errors.
    if (typeof error !== 'object' || error === null
        || Object.prototype.toString.call(error) !== '[object Error]'
        || !Object.hasOwn(error, 'payload')) throw error;
    const value: unknown = (error as Error & { payload: unknown }).payload;
    if (typeof value !== 'object' || value === null || !('tag' in value)
        || typeof value.tag !== 'string' || Object.keys(value).length !== 1
        || !tags.includes(value.tag)) throw error;
    return { tag: 'err', val: value as E };
  }
}

/** Propagate a declared error through a generated WIT result export. */
export function unwrap<T, E>(result: Result<T, E>): T {
  if (result.tag === 'err') throw result.val;
  return result.val;
}
